### PR TITLE
User Story #76 - Content Based Filtering

### PR DESCRIPTION
**Pull request by vikadia7x**
Task #85 - Data Preprocessing Step for model - Created a python notebook file to preprocess metadata - metaDataPreprocessing.ipynb - Data inserted on the central DB
Task #86 - Content Based Filtering using Description - Used Description and keyword to create a content based filtering method - Content_Based.ipynb
Task #87 - Content Based Filtering using all attributes - Used Cast, Crew, Genre, Description, Keywords and Taglines to create a content based filtering engine - Content_Based.ipynb
Task #88 - Model Testing - Created the final recommendation engine py file - content_filtering.py - tested for 5 movies

### CODE REVIEW

File: myProject/Recommendation/content_filtering.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
+from nltk.stem.snowball import SnowballStemmer
+from nltk.stem.wordnet import WordNetLemmatizer
+from nltk.corpus import wordnet
+from surprise import Reader, Dataset, SVD, evaluate
+import warnings; warnings.simplefilter('ignore')
+import pyodbc
+
+"""
+server = 'showtimefinder.database.windows.net'
```
Comment (ShivamVerma920302) [inline myProject/Recommendation/content_filtering.py]: The server details and passwords should not be committed in git. Let's use AzureKeyVault or something similar to store the server details and password. For now let's remove this commented code.

File: myProject/Recommendation/content_filtering.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    metaDataLinks['cast'] = metaDataLinks['cast'].apply(literal_eval)
+    metaDataLinks['crew'] = metaDataLinks['crew'].apply(literal_eval)
+    metaDataLinks['keywords'] = metaDataLinks['keywords'].apply(literal_eval)
+    metaDataLinks['cast_size'] = metaDataLinks['cast'].apply(lambda x: len(x))
+    metaDataLinks['crew_size'] = metaDataLinks['crew'].apply(lambda x: len(x))
+    metaDataLinks['director'] = metaDataLinks['crew'].apply(get_director)
+    metaDataLinks['cast'] = metaDataLinks['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)
+    metaDataLinks['director'] = metaDataLinks['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
+    metaDataLinks['director'] = metaDataLinks['director'].apply(lambda x: [x,x, x])
+    print("Breakpoint")
```
Comment (ShivamVerma920302) [inline myProject/Recommendation/content_filtering.py]: Remove print statement from the code.

File: myProject/Recommendation/content_filtering.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    movie_indices = [i[0] for i in sim_scores]
+    movies = metaDataLinks.iloc[movie_indices][['title','flag','id','imdb_id','year','popularity']]
+    qualified = movies[movies.flag == 1]
+    return qualified
+
+
+def flow():
+    recommendation = pd.DataFrame()
+    metaDataLinks = preProcess(credits,keywords,links,metaData,ratings)
+    print("2")
```
Comment (ShivamVerma920302) [inline myProject/Recommendation/content_filtering.py]: Remove print statements from the code. If required let's implement something like a logger to generate logs in a text file with timestamps.

File: myProject/Recommendation/content_filtering.py (hunk 1 above)
Comment (vikadia7x) [reply myProject/Recommendation/content_filtering.py]: Resolved.

File: myProject/Recommendation/content_filtering.py (hunk 2 above)
Comment (vikadia7x) [reply myProject/Recommendation/content_filtering.py]: Resolved.

File: myProject/Recommendation/content_filtering.py (hunk 3 above)
Comment (vikadia7x) [reply myProject/Recommendation/content_filtering.py]: Resolved.

File: myProject/Recommendation/content_filtering.py (hunk 1 above)
Comment (vikadia7x) [reply myProject/Recommendation/content_filtering.py]: Will take this in Sprint 3

**Comment by ShivamVerma920302** [comment]
Comments are resolved and the code looks good. Merging branch to develop.

**Pull request merged**